Fixes for stdio mode (tested with lsp-mode for emacs)

klajo opened the pull request:
Make DEBUG/TRACE optional at compilation time (off by default).  If
lsp-mode for emacs is run with sourcer in stdio mode (not tcp) then
lsp-mode will fail like on text that is not part of the proper LSP
jsonrpc texts:

    error in process filter: progn: Error parsing language server output: (error No Content-Length header)

Instead of having these (same for TRACE macros) in the source code:

    -define(DEBUG, true).
    -define(DEBUG, 1).

... use rebar3 profiles to enable DEBUG/TRACE builds.  From README.md:

    Run this to build everything:

        rebar3 compile

    Run this to build everything with `DEBUG` printouts enabled:

        rebar3 as debug compile

    Run this to build everything with `TRACE` printouts enabled:

        rebar3 as trace compile

    Run this to build everything with both `DEBUG` and `TRACE` printouts enabled:

        rebar3 as debug,trace compile

    Please note that these will result in different builds below the
    `_build` directory.

klajo committed:
jsonrpc: Bugfix stdio LSP reader

There were two problems with the stdio reader:

1. io:get_line/1 will not work since (at least) lsp-mode sends data with
   a content-length that ends directly after the trailing "}" in the json
   structure, which means that io:get_line won't get that last "}" and
   try_decode never finds a complete message and the main loop "hangs".

   Perhaps other clients behave differently.

2. io:get_line/1 returns a string, whereas the main loop assumed a
   binary here:

        {data, Data} ->
                %?TRACE("STDIO", []),
                Buf2 = <<Buf/binary, Data/binary>>,

io:get_chars/2 would work, but forces us to read 1 character at
a time which is unnecessarily inefficient.

Open a port to read from standard input instead of using one of the io
built-ins.  Requires the erlang node to be started with parameter
"-noinput".

vladdu commented:
Thank you @klajo! It looks good, will try to merge it as soon as I get some time for it. 

Regarding the debug printouts, I'm considering making them go to stderr instead. Only being able to enable them for all modules or for none is a little too coarse.

klajo committed:
Avoid traces to stdout which confuse LSP clients

If lsp-mode for emacs is run with sourcer in stdio mode (not tcp) then
lsp-mode will fail like on text that is not part of the proper LSP
jsonrpc texts:

    error in process filter: progn: Error parsing language server output: (error No Content-Length header)

Pending another logging solution, comment out the define:s which turn on
debug/trace.

klajo commented:
I force-pushed an update to the branch, where I changed the debug/trace to only comment out the defines.  I.e. this PR now contains commented-out defines and the stdio reader fix.
The original "rebar3 as debug compile" [change is here](https://github.com/klajo/sourcer/tree/rebar3-controlled-debug-printouts).

klajo commented:
Thanks for merging!

vladdu commented:
Sure, I have to thank you more for the fix! Sorry that it took a while, I was on vacation.

klajo commented:
No worries.
